File: 99-use-cases/14-progressive-rollout/highlight_eval.py
"""Evaluate configure-grid-selection-green-highlight for grid highlight color.

In this example, we have a progressive rollout over 15 minutes in five equal stages:
10%, 20%, 40%, 60%, and 100% of users receive the green highlight.
"""

from __future__ import annotations

from ldclient import Context

# LaunchDarkly capability: String flag evaluation (server-side SDK)
# See: https://launchdarkly.com/docs/sdk/features/evaluations

# LaunchDarkly: flag key=configure-grid-selection-green-highlight name="Configure: grid selection green highlight" kind=boolean
# https://app.launchdarkly.com/projects/lunch-marcoly/features/configure-grid-selection-green-highlight

FLAG_HIGHLIGHT = "configure-grid-selection-green-highlight"

VALID_COLORS = frozenset({"yellow", "red", "blue", "green", "purple"})


def build_context(username: str) -> Context:
    """User context keyed by login username."""
    return Context.builder(username).build()


def normalize_highlight_color(raw: object) -> str:
    color = str(raw or "none").strip().lower()
    if color in VALID_COLORS:
        return color
    return "none"


def color_label(highlight_color: str) -> str:
    return f"({highlight_color})" if highlight_color != "none" else "(no-color)"


def build_response(username: str, raw: object) -> dict[str, object]:
    color = normalize_highlight_color(raw)
    return {
        "username": username,
        "flagValue": str(raw if raw is not None else "none"),
        "highlightColor": color,
        "colorLabel": color_label(color),
    }


def evaluate_highlight(client, username: str) -> dict[str, object]:
    context = build_context(username)
    if client is None or not client.is_initialized():
        return build_response(username, "none")
    raw = client.variation(FLAG_HIGHLIGHT, context, "none")
    return build_response(username, raw)

File: 99-use-cases/14-progressive-rollout/highlight-eval.js
/** Evaluate enable-grid-selection-highlight for grid highlight color. */

const LaunchDarkly = require("@launchdarkly/node-server-sdk");

// LaunchDarkly capability: String flag evaluation (server-side SDK)
// See: https://launchdarkly.com/docs/sdk/features/evaluations

// LaunchDarkly: flag key=enable-grid-selection-highlight name="Enable: grid selection highlight" kind=string
// https://app.launchdarkly.com/projects/lunch-marcoly/features/enable-grid-selection-highlight

const FLAG_HIGHLIGHT = "enable-grid-selection-highlight";
const VALID_COLORS = new Set(["yellow", "red", "blue", "green", "purple"]);

function buildContext(username) {
  return { kind: "user", key: username };
}

function normalizeHighlightColor(raw) {
  const color = String(raw || "none")
    .trim()
    .toLowerCase();
  return VALID_COLORS.has(color) ? color : "none";
}

function colorLabel(highlightColor) {
  return highlightColor === "none" ? "(no-color)" : `(${highlightColor})`;
}

function buildResponse(username, raw) {
  const color = normalizeHighlightColor(raw);
  return {
    username,
    flagValue: String(raw ?? "none"),
    highlightColor: color,
    colorLabel: colorLabel(color),
  };
}

async function evaluateHighlight(client, username) {
  if (!client) {
    return buildResponse(username, "none");
  }
  const context = buildContext(username);
  const raw = await client.variation(FLAG_HIGHLIGHT, context, "none");
  return buildResponse(username, raw);
}

module.exports = {
  FLAG_HIGHLIGHT,
  buildContext,
  normalizeHighlightColor,
  colorLabel,
  buildResponse,
  evaluateHighlight,
};
